### PR TITLE
Canary/date input

### PR DESCRIPTION
**Pull request by jd239**
## Summary of the changes
- Added new github actions for Cypress this includes:
  - Running Cypress tests only
  - Updating Cypress visual regression snapshots via CI
  - Uploading visual regression report if there are any errors
  - Updating folders names due to visual regression v2 update
- Threshold helper function which allows a 0 threshold value for CI but an updated threshold for local testing
- Date Input implementation for Canary release
- Jest/Puppeteer to Cypress component tests
- Updated ic-input-component-container for additional 'large' size
- Removal of screenshots from v1 visual regression within canary-web-components

## Related issue
#1364 

## Checklist
- [x] Relevant unit tests and visual regression tests added. 
- [x] Visual testing against Figma component specification completed. 
- [x] All acceptance criteria reviewed and met. 
- [x] Accessibility Insights FastPass performed.
- [x] A11y unit test added and yields no issues.
- [x] A11y plug-in on Storybook yields no issues. 
- [x] Manual screen reader testing performed using NVDA and VoiceOver. 
- [x] Page can be zoomed to 400% with no loss of content. 
- [x] Screen magnifier used with no issues. 
- [x] Text resized to 200% with no loss of content.
- [x] Text spacing increased as per the [WCAG 1.4.12 success criterion](https://www.w3.org/TR/WCAG21/#text-spacing) with no loss of content.
- [x] Browser setting 'prefers reduced motion' tested. No animations or motion visible whilst this setting is on. 
- [x] Windows High Contrast mode tested with no loss of content.
- [x] System light and dark mode tested with no loss of content. 
- [x] Manual keyboard testing for keyboard controls and logical focus order. 
- [x] Min/max content examples tested with no loss of content or overflow. 
- [x] Browser support tested (Chrome, Safari, Firefox and Edge).
- [x] Correct roles used and ARIA attributes used correctly where required. 
- [x] Logical heading structure is maintained, and the HTML elements used for headings can be changed to fit within the wider page structure. 
- [x] All prop combinations work without issue.

### CODE REVIEW

File: packages/canary-react/cypress/screenshots/IcDateInput/IcDateInput.cy.tsx/IcDateInput -- renders (failed) (attempt 2).png
Comment (GCHQ-Developer-299) [inline packages/canary-react/cypress/screenshots/IcDateInput/IcDateInput.cy.tsx/IcDateInput -- renders (failed) (attempt 2).png:1]: Should the screenshots for these failure attempts be included in the commit?

Comment (GCHQ-Developer-299) [reply packages/canary-react/cypress/screenshots/IcDateInput/IcDateInput.cy.tsx/IcDateInput -- renders (failed) (attempt 2).png:1]: Oh i see, they're added in commit 1 and removed in commit 4

File: packages/canary-web-components/src/utils/date-helpers.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+export const isDayFirstFormat = (dateString: string) => {
+  const dayFirstFormat =
+    /^(0[1-9]|[12][0-9]|3[01])(\/|-)(0[1-9]|1[0-2])(\/|-)\d{4}$/;
+  return dayFirstFormat.test(dateString);
+};
+
+export const createDate = (date: string | Date, dateFormat: IcDateFormat) => {
+  let newDate;
+  if (isDateOrEpoch(date)) {
+    if (typeof date === "string") {
```
Comment (GCHQ-Developer-299) [inline packages/canary-web-components/src/utils/date-helpers.ts:18]: dunno how you feel about ternary operators but this if/else could be
```js
newDate = (typeof date === "string") ? new Date(date) : date;
```
i think.

File: packages/canary-web-components/src/components/ic-date-input/ic-date-input.tsx [hunk 2]
```diff
@@ -0,0 +1,1547 @@
...
+    return (
+      Object.prototype.toString.call(event) === this.EVENT_OBJECT_STRING ||
+      Object.prototype.toString.call(event) ===
+        this.KEYBOARD_EVENT_OBJECT_STRING
+    );
+  };
+
+  private setDateValidity = () => {
+    if (!isEmptyString(this.day)) {
+      if (+this.day > this.MAX_DAY) {
```
Comment (GCHQ-Developer-299) [inline packages/canary-web-components/src/components/ic-date-input/ic-date-input.tsx:416]: Couple more opportunities to replace with ternary operators here. this.isValidDay and this.isValidMonth

File: packages/canary-web-components/src/components/ic-date-input/ic-date-input.tsx [hunk 3]
```diff
@@ -0,0 +1,1547 @@
...
+      !regex.test(eventKey) &&
+      !(
+        (event.ctrlKey || event.metaKey) &&
+        (eventKey === "v" || eventKey === "c")
+      )
+    ) {
+      event.preventDefault();
+    }
+
+    if (eventKey === "/" || eventKey === "-" || eventKey === ".") {
```
Comment (GCHQ-Developer-299) [inline packages/canary-web-components/src/components/ic-date-input/ic-date-input.tsx:529]: potentially could swap this for a switch statement where it's like 
```js
switch(eventKey) {
case "/":
case "-":
case ".":
  event.preventDefault();
  this.setPasteInvalidText();
  this.moveToNextInput(input);
case: "arrowright":
case: "arrowleft":
  event.preventDefault();
  ...
```
but i dunno what you think is more readable!

**Review by GCHQ-Developer-299 — APPROVED** [review]
Tested stories with keyboard and screenreader and all seemed functional :)

**Review by GCHQ-Developer-299 — APPROVED** [review]

**Review by GCHQ-Developer-465 — APPROVED** [review]

**Review by GCHQ-Developer-299 — APPROVED** [review]

**Review by ad9242 — CHANGES_REQUESTED** [review]
Should not be merged until date-picker component is also approved

**Review by GCHQ-Developer-847 — APPROVED** [review]

**Pull request merged**